src/volatile_memory.rs

```diff
@@ -219,13 +235,95 @@ impl<'a> From<&'a mut [u8]> for VolatileSlice<'a, ()> {
 #[repr(C, packed)]
 struct Packed<T>(T);
 
+struct VolatileSliceGuardInner {
```

Comment:
I don't think we need to encapsulate this. Something like 

```rs
impl VolatileSliceGuard {
   fn new_read(...)
   fn new_write(...)
```

should work instead of `VolatileSliceGuard` and `VolatileSliceGuardMut`. However, this does not actually work the way it is intended, because it doesn't "guard": Since there's no lifetime on raw pointers, nothing prevents the guard (and thus the underlying MmapSlice) from being dropped while we still hold on to the return value of `as_ptr`. I don't think there is a way to properly express this, so if this isn't _needed_ for Xen (which based on our convo in the last meeting it doesn't seem to be), maybe we should offer this? (or just make mmap_slice() public and have people use its `as_ptr` on xen.

Reply:
> However, this does not actually work the way it is intended, because it doesn't "guard": Since there's no lifetime on raw pointers, nothing prevents the guard (and thus the underlying MmapSlice) from being dropped while we still hold on to the return value of `as_ptr`. I don't think there is a way to properly express this, so if this isn't _needed_ for Xen (which based on our convo in the last meeting it doesn't seem to be), maybe we should offer this? (or just make mmap_slice() public and have people use its `as_ptr` on xen.

@roypat when I sketched the guard idea as a hack to test virtiofsd during an earlier iteration (originally it used `MmapXenSlice` and just hacked it public) I mostly thought about it as "guarding the pointer" to be available. Of course nothing is ensuring that the pointer will still be valid, but I guess that is the nature of the pointer? Nothing keeps it valid in case someone just calls .as_ptr() too. But I agree that the name Guard may be misleading. We discussed it this morning and came up with nothing better :(. Also merging with the existing `MmapXenSlice` may be worthwile. However, I wonder: do we really want to fracture the ecosystem by providing two APIs for the same thing?

I understand that it is sensible to keep `.as_ptr()` for backwards compatibility, but only providing the new API through `mmap_slice` (which is Xen specific) would push the conditional compilation into the consumer of this lib. My fear is that nobody will bother and people will only implement one or the other variant and the ecosystem will be split between Xen and KVM.

I think one could maybe rename `MmapXenSlice` to something generic that also exists for KVM (but does nothing really) and expose the mmap() call, but that feels like a weird API for KVM where nothing happens there. @vireshk and me actually discussed calling the guard creation function `mmap` this morning, but rejected that idea because of it feeling weird for KVM.

So maybe we should first discuss which differences between KVM and Xen we want to allow in the API first? Personally, I would suggest to advertise an identical API for both KVM and Xen in order to minimize fragmention. Of course removing .as_ptr() is be too invasive, but we could for example just mark it as deprecated? The extra calls are empty for KVM and should get optimized away, so I do not think that we have any performance issues there.

We are really not aiming at a "on xen" and "on kvm" world here. We want to abstract the differences away as much as possible :).

Reply:
> I don't think we need to encapsulate this. Something like
> 
> ```rust
> impl VolatileSliceGuard {
>    fn new_read(...)
>    fn new_write(...)
> ```

I can't create the Guard without knowing if it is a `read` or `write` operation, as the mapping needs that information. Once I have done the mapping, the only thing I need to return is a pointer, which the consumers can use to avoid an extra copy of data. I am not sure how these new_read/write routines are supposed to work, return. Sorry, a bit confused here :(

> should work instead of `VolatileSliceGuard` and `VolatileSliceGuardMut`. However, this does not actually work the way it is intended, because it doesn't "guard": Since there's no lifetime on raw pointers, nothing prevents the guard (and thus the underlying MmapSlice) from being dropped while we still hold on to the return value of `as_ptr`.

Right. Maybe we need a better name here, as what we are looking for here is a guaranted way by which the memory can be accessed safely.

> I don't think there is a way to properly express this, so if this isn't _needed_ for Xen (which based on our convo in the last meeting it doesn't seem to be), maybe we should offer this? (or just make mmap_slice() public and have people use its `as_ptr` on xen.

We would need this for Xen too, for example we intend to reused Virtiofsd crate, which uses this to avoid extra copying of data. And so a standard way by which they can get the pointer, instead of adding `#cfg(xen)`, would be preferred here.

Reply:
> I understand that it is sensible to keep `.as_ptr()` for backwards compatibility, but only providing the new API through `mmap_slice` (which is Xen specific) would push the conditional compilation into the consumer of this lib. My fear is that nobody will bother and people will only implement one or the other variant and the ecosystem will be split between Xen and KVM.

Also note that we will need something very similar going forward with p-KVM (protected-kvm) as well, as the entire guest memory being made available to the backend won't work for very long. People will definitely want to restrict it. I am not sure of the interface that is going to be available then, but mapping surely must happen on-demand somehow.

Reply:
Right, what I meant with the code change is that the `as_guard` and `as_guard_mut` methods return `VSGuard/VSGuardMut::new` (which then calls VSGuardInner::new with read/write access specified), but instead can just immediately return `VSGuardInner::new_read` and `VSGuardInner::new_write` (or with the current function name, immediately return `VolatileSliceGuardInner::new(..., proto: READ/WRITE)`). No need for the *Mut encapsulation. 

So I was under the impression this wasn't needed for Xen as an earlier iteration of this PR did not include this functionality (as as_ptr was made private and thus not available at all) and I was told that virtiofsd still worked. So in that sense I think it makes more sense to err on the side of less functionality (as opposed to committing something we're unsure of and is not required) and revisit this question at a later point in time (e.g. leave the API unchanged for unix and think about a unification later). I'm open to deprecating as_ptr and finding a generalization, but as a follow up if its not required immediately, to keep this PR minimal wrt to Xen support.

Reply:
> Right, what I meant with the code change is that the `as_guard` and `as_guard_mut` methods return `VSGuard/VSGuardMut::new` (which then calls VSGuardInner::new with read/write access specified), but instead can just immediately return `VSGuardInner::new_read` and `VSGuardInner::new_write` (or with the current function name, immediately return `VolatileSliceGuardInner::new(..., proto: READ/WRITE)`). No need for the *Mut encapsulation.

Ahh I understood it now. I implemented two variants so they can have different `as_ptr()` methods, the read one returning `*const u8` and write one retuning `*mut u*`. If it is fine to return `*mut u8` for both read and write, then I can work with just one structure instead of three. Sure.
 
> So I was under the impression this wasn't needed for Xen as an earlier iteration of this PR did not include this functionality (as as_ptr was made private and thus not available at all) and I was told that virtiofsd still worked.

Maybe I wasn't able to clarify that earlier. Lemme do so now. The early version was tested with vhost-device implementations, which didn't use `as_ptr()` and worked. Then @Ablu tested virtiofsd and it didn't work. We hacked it up a bit to see if Xen grant is working or not and at the same time I privatized `as_ptr()` to make sure it is clear that we can't use it anymore.

In short, we still need something like `as_ptr()` for virtiofsd to work.

> So in that sense I think it makes more sense to err on the side of less functionality (as opposed to committing something we're unsure of and is not required) and revisit this question at a later point in time (e.g. leave the API unchanged for unix and think about a unification later). I'm open to deprecating as_ptr and finding a generalization, but as a follow up if its not required immediately, to keep this PR minimal wrt to Xen support.

As i explained above, it is required right away for virtiofsd. Sorry for the confusion.